### ISSUE TITLE
myopic energy balance unbalanced

### ISSUE THREAD

**Issue by AlexanderMeisinger** [PyPSA/pypsa-eur#1460]
## Checklist

- [x] I am using the current [`master`](https://github.com/PyPSA/pypsa-eur/tree/master) branch or the latest [release](https://github.com/PyPSA/pypsa-eur/releases). Please indicate.
- [x] I am running on an up-to-date [`pypsa-eur` environment](https://github.com/PyPSA/pypsa-eur/blob/master/envs/environment.yaml). Update via `conda env update -f envs/environment.yaml`.

## Describe the Bug
I am running PyPSA-Eur with myopic from 2025 to 2050 (5 years steps). Looking at the results, I noticed that supply and demand are unbalanced.

The attached file was generated with PyPSA-Eur (results/graphs/energy.svg and balances-H2.svg). Config changes compared to default config:
- foresight: myopic
- sectors_opts: 144H-T-H-B-I-A-solar+p3-dist1
- planning_horizons: 2025, 2030, 2035, 2040, 2045, 2050
- co2_budget: 2050: 0.032 instead of 0 - otherwise infeasible optimization problem

The optimization problem is feasible. Therefore, supply and demand has to be balanced. However, it is not shown in the output files.

![balances-H2](https://github.com/user-attachments/assets/4129c2ff-7f5e-4bf3-b1fa-76bf8c34901d)
![energy](https://github.com/user-attachments/assets/d0e04be0-1685-4b87-8dca-f004991da1c1)

**Comment by FabianHofmann** [comment]
yes, this is indeed off. I fear it is because hard-coded fields in the current balance calculations... we definitely have to push #921

**Comment by FabianHofmann** [comment]
On the other hand it looks still very suspicious. Could you make sure, that we are not missing any h2 producing techs (staying with the first example). Also perhaps you could apply 'n.statistics.energy_balance(bus_carrier="h2")' on the network and see whether the error also occurs there

**Comment by AlexanderMeisinger** [comment]
The main h2 producing techs SMR, SMR CC and H2 Electrolysis are  and also have some values, but very small values - not enough to compensate the h2 demand. n.statistics.energy_balance(bus_carrier="h2") shows the same unbalanced results compare to the graphs. I also started the same myopic optimization from 2030 till 2050 (5 years steps) with different unbalanced results:
![balances-H2](https://github.com/user-attachments/assets/fba73f06-48f7-47fc-94b5-63e13e86ef99)
![balances-low_voltage](https://github.com/user-attachments/assets/44be67e3-7f8b-4a92-b2b6-2fdc2dd57694)
![energy](https://github.com/user-attachments/assets/fcc83436-e45f-4849-8844-ebabc8a6ed32)

**Comment by FabianHofmann** [comment]
hey @AlexanderMeisinger , could you make sure you have the latest linopy version installed?

**Comment by FabianHofmann** [comment]
there was a problem with the balance constraint and the new slicing scheme that we introduced. but it was fixed in v4.3

**Comment by AlexanderMeisinger** [comment]
I already checked the linopy package (linopy=0.4.3=pyhd8ed1ab_0). I think, this should be fine. 

[pypsa-eur-packages.txt](https://github.com/user-attachments/files/18210714/pypsa-eur-packages.txt)

**Comment by FabianHofmann** [comment]
thanks, could you do me one more favor and send config updates in yaml form so I can reproduce the error quickly?

**Comment by AlexanderMeisinger** [comment]
Sure, thanks for your support!
[config.pathways-myopics_default.txt](https://github.com/user-attachments/files/18210925/config.pathways-myopics_default.txt)

**Comment by FabianHofmann** [comment]
@AlexanderMeisinger I think I found it. thanks again for bringing it up. this has been under the radar for too long. there was a problem with iterating over constraint slices when writing the problem to file. Under certain circumstances, the last slice of a constraint could be left out (!), which is really bad because it would just be ignored. This is fixed in https://github.com/PyPSA/linopy/pull/398.

Could you please update linopy to 0.4.4 and test again? 

@thomgeo you also told me you saw this problem. i hope this is fixed now

**Comment by AlexanderMeisinger** [comment]
@FabianHofmann Thanks a lot for your support! Updating linopy to 0.4.4 fixed the issue! Well done :)
